pkg/steps/release/promote.go

```diff
@@ -70,89 +64,129 @@ func (s *promotionStep) run(ctx context.Context) error {
-		return retry.RetryOnConflict(promotionRetry, func() error {
-			is, err := s.dstClient.ImageStreams(s.config.Namespace).Get(ctx, s.config.Name, meta.GetOptions{})
-			if errors.IsNotFound(err) {
-				is, err = s.dstClient.ImageStreams(s.config.Namespace).Create(ctx, &imageapi.ImageStream{
-					ObjectMeta: meta.ObjectMeta{
-						Name:      s.config.Name,
-						Namespace: s.config.Namespace,
-					},
-				}, meta.CreateOptions{})
-			}
-			if err != nil {
-				return fmt.Errorf("could not retrieve target imagestream: %w", err)
-			}
+	imageMirror := getImageMirror(s.config, tags, pipeline)
+	if len(imageMirror) == 0 {
+		log.Println("Nothing to promote, skipping...")
+		return nil
+	}
 
-			for dst, src := range tags {
-				if valid, _ := utils.FindStatusTag(pipeline, src); valid != nil {
-					is.Spec.Tags = append(is.Spec.Tags, imageapi.TagReference{
-						Name: dst,
-						From: valid,
-					})
-				}
-			}
+	if _, err := steps.RunPod(ctx, s.podClient, s.eventClient, getPromotionPod(imageMirror, s.jobSpec.Namespace())); err != nil {
+		return fmt.Errorf("unable to run promotion pod: %w", err)
+	}
+	return nil
+}
 
-			if _, err := s.dstClient.ImageStreams(s.config.Namespace).Update(ctx, is, meta.UpdateOptions{}); err != nil {
-				if errors.IsConflict(err) {
-					return err
-				}
-				return fmt.Errorf("could not promote image streams: %w", err)
+func getImageMirror(config api.PromotionConfiguration, tags map[string]string, pipeline *imageapi.ImageStream) map[string]string {
+	if pipeline == nil {
+		return nil
+	}
+	imageMirror := map[string]string{}
+	if len(config.Name) > 0 {
+		for dst, src := range tags {
+			dockerImageReference := findDockerImageReference(pipeline, src)
+			if dockerImageReference == "" {
+				continue
+			}
+			dockerImageReference = getPublicImageReference(dockerImageReference, pipeline.Status.PublicDockerImageRepository)
+			imageMirror[dockerImageReference] = fmt.Sprintf("%s/%s/%s:%s", api.DomainForService(api.ServiceRegistry), config.Namespace, config.Name, dst)
+		}
+	} else {
+		for dst, src := range tags {
+			dockerImageReference := findDockerImageReference(pipeline, src)
+			if dockerImageReference == "" {
+				continue
 			}
-			return nil
-		})
+			dockerImageReference = getPublicImageReference(dockerImageReference, pipeline.Status.PublicDockerImageRepository)
+			imageMirror[dockerImageReference] = fmt.Sprintf("%s/%s/%s:%s", api.DomainForService(api.ServiceRegistry), config.Namespace, dst, config.Tag)
+		}
+	}
+	if len(imageMirror) == 0 {
+		return nil
 	}
+	return imageMirror
+}
 
-	client := s.dstClient.ImageStreamTags(s.config.Namespace)
-	for dst, src := range tags {
-		valid, _ := utils.FindStatusTag(pipeline, src)
-		if valid == nil {
-			continue
-		}
+func getPublicImageReference(dockerImageReference, publicDockerImageRepository string) string {
+	if !strings.Contains(dockerImageReference, ":5000") {
+		return dockerImageReference
+	}
+	i := strings.Index(publicDockerImageRepository, "/")
```

Comment:
This would be more readable with `strings.Split(dockerImageReference, "/")`, then checking the `len` of the result

Reply:
Updated.